src/Nethermind/Nethermind.Consensus/Processing/BlockchainProcessor.cs

```diff
@@ -414,34 +415,30 @@ private void FireProcessingQueueEmpty()
             if (_logger.IsDebug) _logger.Debug($"Skipped processing of {suggestedBlock.ToString(Block.Format.FullHashAndNumber)}, last processed is null: {true}, processedBlocks.Length: {processedBlocks.Length}");
         }
 
+        bool readonlyChain = options.ContainsFlag(ProcessingOptions.ReadOnlyChain);
+        if (!readonlyChain)
+        {
+            long blockProcessingTimeInMicrosecs = _stopwatch.ElapsedMicroseconds();
+            Metrics.LastBlockProcessingTimeInMs = blockProcessingTimeInMicrosecs / 1000;
+            Metrics.RecoveryQueueSize = _recoveryQueue.Count;
+            Metrics.ProcessingQueueSize = _blockQueue.Count;
+            _stats.UpdateStats(lastProcessed, blockProcessingTimeInMicrosecs);
+        }
+
         bool updateHead = !options.ContainsFlag(ProcessingOptions.DoNotUpdateHead);
         if (updateHead)
         {
             if (_logger.IsTrace) _logger.Trace($"Updating main chain: {lastProcessed}, blocks count: {processedBlocks.Length}");
             _blockTree.UpdateMainChain(processingBranch.Blocks, true);
         }
 
-        bool readonlyChain = options.ContainsFlag(ProcessingOptions.ReadOnlyChain);
-        long blockProcessingTimeInMs = _stopwatch.ElapsedMilliseconds;
-        if (!readonlyChain)
-        {
-            Metrics.LastBlockProcessingTimeInMs = blockProcessingTimeInMs;
-        }
-
         if ((options & ProcessingOptions.MarkAsProcessed) == ProcessingOptions.MarkAsProcessed)
         {
             if (_logger.IsTrace) _logger.Trace($"Marked blocks as processed {lastProcessed}, blocks count: {processedBlocks.Length}");
             _blockTree.MarkChainAsProcessed(processingBranch.Blocks);
-
-            Metrics.LastBlockProcessingTimeInMs = blockProcessingTimeInMs;
         }
 
-        if (!readonlyChain)
-        {
-            Metrics.RecoveryQueueSize = _recoveryQueue.Count;
-            Metrics.ProcessingQueueSize = _blockQueue.Count;
-            _stats.UpdateStats(lastProcessed, _blockTree, _stopwatch.ElapsedMicroseconds());
-        }
+        Metrics.BestKnownBlockNumber = _blockTree.BestKnownNumber;
```

Comment:
I think this should too be under the `if (!readonlyChain)`

Reply:
Fixed